azure/src/main/scala/quasar/blobstore/azure/Azure.scala

```diff
@@ -17,43 +17,98 @@
 
-  def mkCredentials(cred: Option[AzureCredentials]): ICredentials =
+  def mkCredentials[F[_]: ConcurrentEffect](cred: Option[AzureCredentials]): F[Expires[ICredentials]] =
     cred match {
-      case None => new AnonymousCredentials
-      case Some(c) => new SharedKeyCredentials(c.accountName.value, c.accountKey.value)
+      case None =>
+        Expires.never(
+          new AnonymousCredentials: ICredentials).pure[F]
+
+      case Some(AzureCredentials.SharedKey(accountName, accountKey)) =>
+        Expires.never(
+          new SharedKeyCredentials(accountName.value, accountKey.value): ICredentials).pure[F]
+
+      case Some(AzureCredentials.ActiveDirectory(clientId, tenantId, clientSecret)) =>
+        rx.publisherToStream[F, AccessToken](
+          (new ClientSecretCredentialBuilder)
+            .clientId(clientId.value)
+            .tenantId(tenantId.value)
+            .clientSecret(clientSecret.value)
+            .build()
+            .getToken((new TokenRequestContext)
+              .setScopes(List("https://storage.azure.com/.default").asJava)))
+          .compile
+          .lastOrError
+          .map(tk => Expires(new TokenCredentials(tk.getToken), tk.getExpiresAt))
     }
 
-  def mkContainerUrl[F[_]](cfg: Config)(implicit F: Sync[F]): F[ContainerURL] =
-    F.catchNonFatal(new URL(cfg.storageUrl.value)) map { url =>
-      val serviceUrl = new ServiceURL(
-        url,
-        StorageURL.createPipeline(
-          mkCredentials(cfg.credentials),
-          // Azure SDK changed its logging behavior in 10.3.0 (compared to 10.2.0).
-          // In 10.2.0 the standard `new PipelineOptions` could be used, but in 10.3.0
-          // this results in messages being logged to console in eg quasar's repl
-          // (despite having no console logger configured).
-          // Still not having a full understanding of how this logging is supposed to work
-          // exactly (e.g. `withLogger`) but adding `LoggingOptions` with
-          // `disableDefaultLogging = true` seems to be fixing it.
-          // As an aside, in case we need to look at this in the future,
-          // Azure SDK has 2 loggers for some reason:
-          // - `com.microsoft.azure.storage.blob.LoggingFactory`
-          // - `Azure Storage Java SDK`
-          new PipelineOptions().withLoggingOptions(
-            new LoggingOptions(3000, true))))
-      serviceUrl.createContainerURL(cfg.containerName.value)
+  def mkContainerUrl[F[_]](cfg: Config)(implicit F: ConcurrentEffect[F]): F[Expires[ContainerURL]] =
+    F.catchNonFatal(new URL(cfg.storageUrl.value)) flatMap { url =>
+      // Azure SDK changed its logging behavior in 10.3.0 (compared to 10.2.0).
+      // In 10.2.0 the standard `new PipelineOptions` could be used, but in 10.3.0
+      // this results in messages being logged to console in eg quasar's repl
+      // (despite having no console logger configured).
+      // exactly (e.g. `withLogger`) but adding `LoggingOptions` with
```

Comment:
Looks like comment line above `exactly ...` is lost

Reply:
Oh snap. Will fix.